kore/src/Kore/Step/Step.hs

```diff
@@ -438,7 +448,61 @@ recoveryFunctionLikeResults
     => Pattern (Target variable)
     -> unifier (Results variable)
     -> unifier (Results variable)
-recoveryFunctionLikeResults _ = id
+recoveryFunctionLikeResults initial unifier = do
+    results <- Results.results <$> unifier
+    let appliedRules = Result.appliedRule <$> results
+    case checkFunctionLike appliedRules (Conditional.term initial) of
+        Right _  -> unifier
+        Left err -> moreChecksAfterError appliedRules err
+  where
+    moreChecksAfterError appliedRules err = do
+        let
+            appliedRule = case Seq.length appliedRules of
+                1 -> Seq.index appliedRules 0
+                _ -> error $ show $ Pretty.vsep
+                        [ "Expected singleton list of rules but found: "
+                        , (Pretty.indent 4 . Pretty.pretty . show) appliedRules
+                        , "This should be imposssible, as simplifiers for \
+                          \simplification are built from a single rule."
+                        ]
+
+            Conditional { term = ruleTerm, substitution = ruleSubstitution } =
+                appliedRule
+            RulePattern { left = alpha_t', requires = alpha_p'} = ruleTerm
+
+            substitution' = Substitution.toMap ruleSubstitution
+
+            alpha_t = TermLike.substitute substitution' alpha_t'
+            alpha_p = Predicate.substitute substitution' alpha_p'
+            phi_p = Conditional.predicate initial
+            phi_t = Conditional.term initial
+
+        res1 <- SMT.Evaluator.evaluate
+                    $ Predicate.makeAndPredicate
+                        phi_p
+                        (Predicate.makeNotPredicate alpha_p)
+
+        Monad.when (res1 /= Just False) $ error $ show $ Pretty.vsep
+            [ "Couldn't recover simplification coverage error: "
+            , Pretty.indent 4 (Pretty.pretty err)
+            , "Configuration condition "
+            , Pretty.indent 4 $ unparse phi_p
+            , "doesn't imply rule condition "
+            , Pretty.indent 4 $ unparse alpha_p
+            ]
+        let alpha_t_sorted = fullyOverrideSort' (termLikeSort phi_t) alpha_t
+        Monad.when (phi_t /= alpha_t_sorted) $ error $ show $ Pretty.vsep
```

Comment:
Could you add a comment here saying that what we would like to check is that `\implies(phi_t  alpha_t_sorted)`, but that's hard to do for non-functional patterns, so we check for (syntactic) equality instead?

BTW, the design document says "This could be checked with an SMT", but I'm not sure how to do that.

Reply:
> BTW, the design document says "This could be checked with an SMT", but I'm not sure how to do that.

With a very smart SMT which would know all about equality in Matching Logic :-)